internal/gps/checks.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"github.com/golang/dep/internal/gps/pkgtree"
+)
+
+func FindIneffectualConstraints(manifest Manifest, packageTree pkgtree.PackageTree) []ProjectRoot {
+	var stdLibFn func(string) bool
+	// will track return value here
+	ineffectuals := make([]ProjectRoot, 0, 10)
+
+	// flatten list of actual imports which should be checked against
+	reachmap, _ := packageTree.ToReachMap(true, true, false, nil /*ignores*/)
+	reach := reachmap.FlattenFn(stdLibFn)
+	imports := make(map[string]bool)
+	for _, imp := range reach {
+		imports[imp] = true
+	}
+
+	// if manifest is a RootManifest, use requires and ignores
+	if rootManifest := manifest.(RootManifest); rootManifest != nil {
+		// add required packages
+		for projectRoot, _ := range rootManifest.RequiredPackages() {
+			imports[projectRoot] = true
+		}
+
+		// check that ignores actually refer to packages we're importing
+		for ignore, _ := range rootManifest.IgnoredPackages() {
```

Comment:
We can't make this check authoritative, as ignores are applied globally (to both direct and transitively imported packages). The pkgtree only tells us about the first layer of externally imported packages, so we can't know if an ignore that isn't in that set is potentially in the transitive set.

We CAN, though, check and see if there are ignore statements for any _local_ project packages that don't exist. That is, if `pkgtree.ImportRoot` is `github.com/foo/bar`, and there's a directive to ignore `github.com/foo/bar/baz`, but that package does not exist in the package tree, then we can indicate that it's an ineffectual ignore.